contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1049 @@
+  );
+
+  // todo document expected initialzation state (this is a holdover from LockedNORI, not totally sure what it means)
+  function initialize() external initializer {
+    __ERC1155_init_unchained(
+      "https://nori.com/api/restrictionschedule/{id}.json"
+    );
+    __Context_init_unchained();
+    __ERC165_init_unchained();
+    __AccessControl_init_unchained();
+    __AccessControlEnumerable_init_unchained();
+    __Pausable_init_unchained();
+    __ERC1155Supply_init_unchained();
+    _setupRole(DEFAULT_ADMIN_ROLE, _msgSender());
+    _setupRole(PAUSER_ROLE, _msgSender());
+    _setupRole(SCHEDULE_CREATOR_ROLE, _msgSender());
+    _setupRole(TOKEN_REVOKER_ROLE, _msgSender());
+    // Seconds in 10 years, based on average year duration of 365.2425 days, which accounts for leap years
+    setRestrictionDurationForMethodologyAndVersion(1, 0, 315_569_520);
+    _ERC1820_REGISTRY.setInterfaceImplementer(
+      address(this),
+      ERC777_TOKENS_RECIPIENT_HASH,
+      address(this)
+    );
+  }
+
+  // View functions and getters =========================================
+  function supportsInterface(bytes4 interfaceId)
+    public
+    view
+    override(ERC1155Upgradeable, AccessControlEnumerableUpgradeable)
+    returns (bool)
+  {
+    return super.supportsInterface(interfaceId);
+  }
+
+  /**
+   * Returns the schedule duration in seconds that has been set for a given methodology and
+   * methodology version.
+   */
+  function getRestrictionDurationForMethodologyAndVersion(
+    uint256 methodology,
+    uint256 methodologyVersion
+  ) public view returns (RestrictionDuration memory) {
+    return
+      _methodologyAndVersionToScheduleDuration[methodology][methodologyVersion];
+  }
+
+  /** Returns an array of all existing schedule ids, regardless of the status of the schedule. */
+  function getAllScheduleIds() external view returns (uint256[] memory) {
+    uint256[] memory allScheduleIdsArray = new uint256[](
+      _allScheduleIds.length()
+    );
+    for (uint256 i = 0; i < allScheduleIdsArray.length; i++) {
+      allScheduleIdsArray[i] = _allScheduleIds.at(i);
+    }
+    return allScheduleIdsArray;
+  }
+
+  /** Returns an array of all schedule ids of which an address currently owns any tokens. */
+  function getScheduleIdsForAccount(address account)
+    external
+    view
+    returns (uint256[] memory)
+  {
+    EnumerableSetUpgradeable.UintSet
+      storage scheduleIdSet = _addressToScheduleIdSet[account];
+    return _enumerableSetToArray(scheduleIdSet);
+  }
+
+  /** Returns an account-specific view of the details of a specific schedule. */
+  function getScheduleDetailForAccount(address account, uint256 scheduleId)
+    public
+    view
+    returns (ScheduleDetailForAddress memory)
+  {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+
+    return
+      ScheduleDetailForAddress(
+        account,
+        scheduleId,
+        schedule.startTime,
+        schedule.endTime,
```

Comment:
seems weird to me to return this data as part of the schedule summary considering that you can get this information independently using the scheduleId parameter. Causes a lot of duplication in data across structs

Reply:
also, is it not safe to assume that if `Schedule.balance > 0` that it exists?

Reply:
sure, the start time and end time? will get rid of

Reply:
oh I didn't see your second comment because this hadn't refreshed.  no that's not a safe assumption to make because schedules get created before they are necessarily funded with anything

Reply:
1. Would also get rid of the scheduleId as that is already known since it's given via the params
2. Instead of using balance you could probably just check that `.account` isn't the zero address then

Reply:
I don't want to get rid of the schedule id because when you call `batchGetScheduleDetailsForAccount` you get an array of schedule details back but you didn't pass in an array of ids, so that helps you know which is which.  otherwise you'd have to make a second call to get an array of schedule ids for an account and things could change between the two calls so you ca'nt just rely on index alignment

Reply:
I don't understand comment #2

Reply:
I think that makes sense about returning the scheduleId, didnt catch that in the batch variant (though that function should probably accept an array of schedule ids as well tbh otherwise external implementers might sooner run out of gas with accounts with many schedules)
comment 2: It seems like the exists property of the struct is not needed as a check for `tokenHolder !== address(0)` should be sufficient

Reply:
the actual schedule struct in storage looks like this: 
```
  struct ScheduleSummary {
    uint256 scheduleTokenId;
    uint256 startTime;
    uint256 endTime;
    uint256 totalSupply;
    uint256 totalClaimableAmount;
    uint256 totalClaimedAmount;
    uint256 totalQuantityRevoked;
    address[] tokenHolders;
    bool exists;
  }
```

So there isn't actually a single `address` field on here that we can check for being the zero address (since a schedule can have multiple holder addresses).  maybe we could check if the length of `tokenHolders` is greater than 0? or if the `scheduleTokenId` is 0? But I feel like we take a hit on readability without gaining much

Reply:
That's the wrong struct, no? 
```sol
  struct ScheduleDetailForAddress {
    address tokenHolder;
    uint256 scheduleTokenId;
    uint256 balance;
    uint256 claimableAmount;
    uint256 claimedAmount;
    uint256 quantityRevoked;
    bool exists;
  }
```

Definitely disagree there. You can almost always imply `exists` by looking at the return data of a struct.

`exists` was primarily introduced in lNORI to help with diffs in the vesting task vs github grant data.

We carried that pattern over to this contract, but it doesn't seem very useful in any place I can identify and further adds one additional value to track and update correctly (e.g., structs who rely on explicitly setting exists = true now also need to always remember to explicitly set  exists = false when wanting to remove the data) **There are some exceptions to this rule, but not sure if any of these structs contain one**

Reply:
so it sounds like I should get rid of `exists` both structs

Reply:
ah ok, the view struct.  it sounds like you're only necessarily wanting to get rid of `exists` on this `ScheduleDetailForAddress` struct then? (based also on the suggestions below).  Can do. let me know if you also want to get rid of the `exists` on the `Schedule` struct itself too

Reply:
I'd love to get rid of the pattern universally. I feel like the reason it's needed in most cases can be handled otherwise. In the cases where we can't there's probably something wrong with our underlying data patterns

Reply:
ok, I'm going to leave this as a todo on this PR and we can decide what to do about it globally.  the view struct has it removed but the struct that's used for logic in the contract i'm going to leave it there for now with a todo